src/middleware/upload.ml

```diff
@@ -7,209 +7,118 @@
 
 module Dream = Dream__pure.Inmost
 
-
-
-let chunk_callback (state : Dream.multipart_state) ~name ~filename data =
-  let on_continue, continue = Lwt.wait () in
-  state.continue <- continue;
-
-  begin match state.event_listener with
-  | Some event_listener ->
-    state.event_listener <- None;
-    state.last_field_name <- Some name;
-    state.last_file_name <- Some filename;
-    state.buffered_chunk <- Some data;
-    Lwt.wakeup_later event_listener (`File (name, filename))
-
-  | None ->
-    match state.chunk_listener with
-    | Some chunk_listener ->
-      state.chunk_listener <- None;
-      if state.last_field_name = Some name &&
-         state.last_file_name = Some filename then
-        Lwt.wakeup_later chunk_listener (Some data)
-      else begin
-        state.last_field_name <- Some name;
-        state.last_file_name <- Some filename;
-        state.buffered_chunk <- Some data;
-        state.next_file <- true;
-        Lwt.wakeup_later chunk_listener None
-      end
-
+let field_to_string (request : Dream.request) field =
+  let open Multipart_form in
+  match field with
+  | Field.Field (field_name, Field.Content_type, v) ->
+    Lwt.return (`Field ((field_name :> string), Content_type.to_string v))
+  | Field.Field (field_name, Field.Content_disposition, v) ->
+    request.specific.upload.filename <- Content_disposition.filename v ;
+    request.specific.upload.name <- Content_disposition.name v ;
+    Lwt.return (`Field ((field_name :> string), Content_disposition.to_string v))
+  | Field.Field (field_name, Field.Content_encoding, v) ->
+    Lwt.return (`Field ((field_name :> string), Content_encoding.to_string v))
+  | Field.Field (field_name, Field.Field, v) ->
+    Lwt.return (`Field ((field_name :> string), Unstrctrd.to_utf_8_string v))
```

Comment:
AFAICT this changes the meaning of `` `Field`` from "non-file form field" (e.g. corresponding to `<input type="text">`) to "header field`, (e.g. `Content-Disposition: ...`). It looks like, in at least some cases, it is redundant. For example, the code already appears to retrieve the file name and field name.

How does `multipart-form-data` distinguish non-file form fields and file fields? Reading RFC 7578, I see no reliable way to do it. Is there some non-RFC convention that has developed in practice in HTTP?

I want to note that even though the Dream-level API has almost the same constructors as the previous `mutlipart-form-data`-based API, it appears to use them completely differently.

If there is no reliable way to distinguish non-files from files at the protocol level (in practice, regardless of RFC), then it seems that both the `multipart-form-data` and the Dream APIs are wrong. They probably need a fatter `part` type that carries all the headers sent with each part, and can be passed to the application to interpret according to its own knoweldge of the meaning of each part. Is that what you were referring to with:

> I'm not very convince by the API

> I don't have the time to figure out about a better interface

I'll do it in follow-on work, but please do let me know what the problems of the current interface are (without necessarily figuring out a new one :P)

Reply:
From my point of view, a `multipart/form-data` (according to RFCs) is a tree of parts (leafs) where you can find for each of them several HTTP fields (such as `Content-Disposition` or `Content-Transfer-Encoding`). So the current behavior of:
```ocaml
val upload : request -> [ Field of string * string | Part of (string option * string option ]
```

Is to deliver to the end-user fields of the the current part as a _stream_ such as:
```ocaml
# upload request ;;
- [> `Field of string * string ] = `Field ("content-type", "text/html")
# upload request ;;
- [> `Field of string * string ] = `Field ("content-disposition", ...)
# upload request ;;
- [> `Field of string * string ] = `Field ("content-transfer-encoding", "base64")
# upload request ;;
- [> `Part of string option * string option ] = `Part (possible_name, possible_filename)
# upload_part request ;;
# upload request ;;
- [> `Field of string * string ] = `Field ("content-type", "image/png")
...
```

It's like a stream of flattened parts. So we don't dissociate files and values. We just say that we have a part with a possible `name` and a possible `filename` extracted from a possible `Content-Disposition`. Note that the part can contains several others fields (like `Comment` or extra information). Even if I did not notice that, it is allowed by the RFC.